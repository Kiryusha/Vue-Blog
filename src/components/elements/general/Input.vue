<template lang="pug">
  .input(
    :class="['input', {'_valid': valid && !error, '_error': error}, cssClass]"
  )
    input.control(
      :placeholder="placeholder"
      :name="name"
      :type="type"
      :disabled="disabled"
      :maxlength="maxlength"
      v-model="innerValue"
      @input="inputHandler"
      @blur="$emit('blur')"
      @focus="$emit('focus')"
    )
    .error(v-if="error") {{ error }}
</template>

<script>
export default {
  props: {
    type: {
      type: String,
      default: 'text',
    },
    placeholder: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      default: '',
    },
    maxlength: {
      type: Number,
      default: 200,
    },
    error: {
      type: String,
      default: '',
    },
    value: {
      type: String,
      default: '',
    },
    cssClass: {
      type: String,
      default: '',
    },
    valid: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      innerValue: this.value,
    };
  },
  watch: {
    value(value) {
      this.innerValue = value;
    },
  },
  methods: {
    inputHandler() {
      this.$emit('input', this.innerValue);
    },
  },
};
</script>

<style lang="stylus" scoped>
  .input
    box-sizing border-box
    -webkit-tap-highlight-color rgba(0, 0, 0, 0)

    &._error .control
      border-color red

  .control
    width 100%
    font-size 16px
    line-height 1.5
    border 1px solid rgba(#000, .15)
    border-radius 4px
    padding 6px 10px
    -webkit-appearance none
    display block
    font-family inherit
    outline none
    resize none
    overflow hidden
    word-wrap break-word

  .error
    font-size 14px
    color red
    padding-top 10px

</style>
