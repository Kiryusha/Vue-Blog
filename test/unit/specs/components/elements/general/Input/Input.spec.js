// Input.vue

import { shallowMount } from '@vue/test-utils';
import Input from 'Components/elements/general/Input.vue';

describe('Input.vue', () => {

  describe('component shallow rendering', () => {
    it('should match a snapshot', () => {
      const wrapper = shallowMount(Input);

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should match a snapshot - with error', () => {
      const wrapper = shallowMount(Input);

      wrapper.setProps({
        error: 'test error',
      });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('custom actions', () => {
    it('should set innerValue on value prop change', () => {
      const wrapper = shallowMount(Input);

      wrapper.setProps({
        value: 'test',
      });

      expect(wrapper.vm.innerValue).toBe('test');
    });

    it('should emit parent input event on local input event', () => {
      const wrapper = shallowMount(Input);

      wrapper.find('.control').trigger('input');

      expect(wrapper.emitted().input).toBeTruthy();
    });
  });
});
