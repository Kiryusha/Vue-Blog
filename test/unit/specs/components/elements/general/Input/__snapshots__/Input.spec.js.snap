// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.vue component shallow rendering should match a snapshot - with error 1`] = `
<div class="input input _error">
  <input placeholder="" name="" maxlength="200" type="text" class="control">
  <div class="error">test error</div>
</div>
`;

exports[`Input.vue component shallow rendering should match a snapshot 1`] = `
<div class="input input">
  <input placeholder="" name="" maxlength="200" type="text" class="control">
  <!---->
</div>
`;
